hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/subscription/SubscriptionHealthCalculator.java

```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    private EnumSet<Problem> calculateHealthProblems(Subscription subscription, TopicMetrics topicMetrics, SubscriptionMetrics subscriptionMetrics) {
+        EnumSet<Problem> healthProblems = noneOf(Problem.class);
+        if (isLagging(topicMetrics, subscriptionMetrics)) {
+            healthProblems.add(LAGGING);
+        }
+        if (isSlow(topicMetrics, subscriptionMetrics)) {
+            healthProblems.add(SLOW);
+        }
+        if (isUnreachable(subscriptionMetrics)) {
+            healthProblems.add(UNREACHABLE);
+        }
+        if (isTimingOut(subscriptionMetrics)) {
+            healthProblems.add(TIMING_OUT);
+        }
+        if (isMalfunctioning(subscriptionMetrics)) {
+            healthProblems.add(MALFUNCTIONING);
+        }
+        if (isReceivingMalformedMessages(subscription, subscriptionMetrics)) {
+            healthProblems.add(RECEIVING_MALFORMED_MESSAGES);
+        }
+        return healthProblems;
+    }
+
+    private boolean isLagging(TopicMetrics topicMetrics, SubscriptionMetrics subscriptionMetrics) {
+        double lagInSeconds = subscriptionMetrics.lag / topicMetrics.rate;
+        return lagInSeconds > MAX_LAG_IN_SECONDS;
+    }
+
+    private boolean isSlow(TopicMetrics topicMetrics, SubscriptionMetrics subscriptionMetrics) {
+        return subscriptionMetrics.rate < MIN_SUBSCRIPTION_TO_TOPIC_SPEED_RATIO * topicMetrics.rate;
+    }
+
+    private boolean isUnreachable(SubscriptionMetrics subscriptionMetrics) {
+        return subscriptionMetrics.otherErrorsRate > MAX_OTHER_ERRORS_RATIO * subscriptionMetrics.rate;
+    }
+
+    private boolean isTimingOut(SubscriptionMetrics subscriptionMetrics) {
+        return subscriptionMetrics.timeoutsRate > MAX_TIMEOUTS_RATIO * subscriptionMetrics.rate;
+    }
+
+    private boolean isMalfunctioning(SubscriptionMetrics subscriptionMetrics) {
+        return subscriptionMetrics.code5xxErrorsRate > MAX_5XX_ERRORS_RATIO * subscriptionMetrics.rate;
+    }
+
+    private boolean isReceivingMalformedMessages(Subscription subscription, SubscriptionMetrics subscriptionMetrics) {
+        return hasClientErrorRetry(subscription) && isCode4xxErrorRateHigh(subscriptionMetrics);
+    }
+
+    private boolean hasClientErrorRetry(Subscription subscription) {
+        if (subscription.isBatchSubscription()) {
+            return subscription.getBatchSubscriptionPolicy().isRetryClientErrors();
+        } else {
+            return subscription.getSerialSubscriptionPolicy().isRetryClientErrors();
+        }
+    }
+
+    private boolean isCode4xxErrorRateHigh(SubscriptionMetrics subscriptionMetrics) {
+        return subscriptionMetrics.code4xxErrorsRate > MAX_4XX_ERRORS_RATIO * subscriptionMetrics.rate;
+    }
+
+    private static final class TopicMetrics {
+        private final double rate;
+
+        private TopicMetrics(pl.allegro.tech.hermes.api.TopicMetrics that) {
+            this.rate = parseDouble(that.getRate());
```

Comment:
There is a null check in JS: https://github.com/allegro/hermes/blob/master/hermes-console/static/js/console/subscription/SubscriptionHealth.js#L6 - I omitted it, is it necessary?